packages/miniflare/src/workers/core/proxy.worker.ts

```diff
@@ -0,0 +1,312 @@
+import {
+  CoreHeaders,
+  ProxyAddresses,
+  ProxyOps,
+  isFetcherFetch,
+  isR2ObjectWriteHttpMetadata,
+} from "./constants";
+import {
+  PlatformImpl,
+  ReducersRevivers,
+  createHTTPReducers,
+  createHTTPRevivers,
+  parseWithReadableStreams,
+  stringifyWithStreams,
+  structuredSerializableReducers,
+  structuredSerializableRevivers,
+} from "./devalue";
+
+const ENCODER = new TextEncoder();
+const DECODER = new TextDecoder();
+
+const WORKERS_PLATFORM_IMPL: PlatformImpl<ReadableStream> = {
+  Blob,
+  File,
+  Headers,
+  Request,
+  Response,
+
+  isReadableStream(value): value is ReadableStream {
+    return value instanceof ReadableStream;
+  },
+  bufferReadableStream(stream) {
+    return new Response(stream).arrayBuffer();
+  },
+  unbufferReadableStream(buffer) {
+    const body = new Response(buffer).body;
+    assert(body !== null);
+    return body;
+  },
+};
+
+interface JsonError {
+  message?: string;
+  name?: string;
+  stack?: string;
+  cause?: JsonError;
+}
+
+function reduceError(e: any): JsonError {
+  return {
+    name: e?.name,
+    message: e?.message ?? String(e),
+    stack: e?.stack,
+    cause: e?.cause === undefined ? undefined : reduceError(e.cause),
+  };
+}
+
+async function readPrefix(
+  stream: ReadableStream<Uint8Array>,
+  prefixLength: number
+): Promise<[prefix: Uint8Array, rest: ReadableStream]> {
+  const reader = await stream.getReader({ mode: "byob" });
+  const result = await reader.readAtLeast(
+    prefixLength,
+    new Uint8Array(prefixLength)
+  );
+  assert(result.value !== undefined);
+  reader.releaseLock();
+  // TODO(cleanup): once https://github.com/cloudflare/workerd/issues/892 fixed,
+  //  should just be able to use `stream` here
+  const rest = stream.pipeThrough(new IdentityTransformStream());
+  return [result.value, rest];
+}
+
+// Helpers taken from `devalue` (unfortunately not exported):
+// https://github.com/Rich-Harris/devalue/blob/50af63e2b2c648f6e6ea29904a14faac25a581fc/src/utils.js#L31-L51
+const objectProtoNames = Object.getOwnPropertyNames(Object.prototype)
+  .sort()
+  .join("\0");
+function isPlainObject(value: unknown) {
+  const proto = Object.getPrototypeOf(value);
+  return (
+    proto === Object.prototype ||
+    proto === null ||
+    Object.getOwnPropertyNames(proto).sort().join("\0") === objectProtoNames
+  );
+}
+function getType(value: unknown) {
+  return Object.prototype.toString.call(value).slice(8, -1); // `[object <type>]`
+}
+
+export class ProxyServer implements DurableObject {
+  // On the first `fetch()`, start a `setInterval()` to keep this Durable Object
+  // and its heap alive. This is required to ensure heap references stay valid
+  // for the lifetime of this `workerd` process (except it isn't since `workerd`
+  // doesn't evict Durable Objects yet :P, but it probably will soon).
+  anchorInterval?: number;
+  nextHeapAddress = ProxyAddresses.USER_START;
+  readonly heap = new Map<number, unknown>();
```

Comment:
Would this work with storing objects in DO storage? Since it accepts v8 serialisable values, it _might_ work? That would allow for removing of the proxy logic